src/arch/MovieTexture/MovieTexture_FFMpeg.cpp

```diff
@@ -164,93 +166,128 @@ float MovieDecoder_FFMpeg::GetTimestamp() const
 	}
 
 	// In a logical situation, this means that display is outpacing decoding.
-	if (display_frame_num_ >= static_cast<int>(frame_buffer_.size())) {
+	if (display_frame_num_ >= static_cast<int>(packet_buffer_.size())) {
+		return 0;
+	}
+
+	PacketHolder* packet = packet_buffer_[display_frame_num_].get();
+
+	// Sanity check.
+	if (packet == nullptr) {
 		return 0;
 	}
-	return frame_buffer_[display_frame_num_]->frame_timestamp;
+
+	std::lock_guard<std::mutex> lock(packet->lock);
+	return packet->frame_timestamp;
 }
 
 bool MovieDecoder_FFMpeg::IsCurrentFrameReady() {
-	// We're displaying faster than decoding. Do not even try to display the frame.
-	if (display_frame_num_ >= static_cast<int>(frame_buffer_.size())) {
+	// We're displaying faster than decoding. Do not even try to display the
+	// frame.
+	if (display_frame_num_ >= static_cast<int>(packet_buffer_.size())) {
 		return false;
 	}
-	// If the whole movie is decoded, then the frame is definitely ready.
-	if (end_of_file_) {
-		return true;
-	}
 
+	FrameHolder* frame = frame_buffer_[(display_frame_num_ + offset_) % frame_buffer_.size()].get();
```

Comment:
same here as above, you could use `std::next` with an iterator and then do `FrameHolder* frame = it->get();`